velox/functions/sparksql/String.h

```diff
@@ -455,4 +455,139 @@ struct SubstrFunction {
+    } else {
+      length = len >= 0 ? len : replace.size();
+    }
+    int64_t start = (int64_t)pos + (int64_t)length;
+    pair = substring<isAscii, isVarchar>(input, start, INT32_MAX);
+    if constexpr (isVarchar) {
+      auto byteRange = stringCore::getByteRange<isAscii>(
+          input.data(), pair.first + 1, pair.second);
+      result.append(StringView(
+          input.data() + byteRange.first, byteRange.second - byteRange.first));
+    } else {
+      result.append(StringView(input.data() + pair.first, pair.second));
+    }
+  }
+
+  // Information regarding the pos calculation:
+  // Hive and SQL use one-based indexing for SUBSTR arguments but also accept
+  // zero and negative indices for start positions. If a start index i is
+  // greater than 0, it refers to element i-1 in the sequence. If a start index
+  // i is less than 0, it refers to the -ith element before the end of the
+  // sequence. If a start index i is 0, it refers to the first element. Return
+  // pair of first indices and length.
+  template <bool isAscii, bool isVarchar>
+  FOLLY_ALWAYS_INLINE std::pair<int32_t, int32_t>
+  substring(const StringView& input, const int64_t pos, const int64_t length) {
+    int64_t len = 0;
+    if constexpr (isVarchar) {
+      len = stringImpl::length<isAscii>(input);
+    } else {
+      len = input.size();
+    }
+    int64_t start = (pos > 0) ? pos - 1 : ((pos < 0) ? len + pos : 0);
+    int64_t end = 0;
+    if (start + length > INT32_MAX) {
+      end = INT32_MAX;
+    } else if (start + length < INT32_MIN) {
+      end = INT32_MIN;
+    } else {
+      end = start + length;
+    }
+
+    if (end <= start || start >= len) {
+      return std::make_pair(0, 0);
+    }
+
+    int64_t zero = 0;
+    int32_t i = std::min(len, std::max(zero, start));
+    int32_t j = std::min(len, std::max(zero, end));
+
+    if (j > i) {
+      return std::make_pair(i, j - i);
+    } else {
+      return std::make_pair(0, 0);
+    }
+  }
+};
+
+template <typename T>
+struct OverlayVarcharFunction : public OverlayFunctionBase {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    OverlayFunctionBase::doCall<false, true>(result, input, replace, pos, len);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    OverlayFunctionBase::doCall<true, true>(result, input, replace, pos, len);
+  }
+};
+
+template <typename T>
+struct OverlayVarbinaryFunction : public OverlayFunctionBase {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
```

Comment:
This applies only to functions that take VARCHAR input and produce VARCHAR result. Let's remoce.